File: user-frontend/src/services/order.ts
import { http } from './http'
import type { Address } from './address'

export type CartItem = {
  sku_id: number
  product_id: number
  merchant_id: number
  merchant_name: string
  merchant_logo_url?: string | null
  product_name: string
  sku_name: string
  price_cent: number
  quantity: number
  checked: boolean
  cover_url?: string | null
  source_post_id?: number | null
  source_label?: string | null
  invalid_reason?: string | null
}

export type CheckoutResult = {
  items: CartItem[]
  addresses: Address[]
  available_full_discounts: Array<{
    id: number
    name: string
    scope_type: string
    scope_ids: number[]
    min_amount_cent: number
    discount_amount_cent: number
    applicable_amount_cent: number
    available: boolean
    unavailable_reason?: string | null
    selected: boolean
  }>
  available_coupons: Array<{
    id: number
    coupon_template_id: number
    name: string
    scope_type: string
    scope_ids: number[]
    discount_type: string
    discount_value: number
    min_amount_cent: number
    applicable_amount_cent: number
    discount_amount_cent: number
    status: string
    available: boolean
    unavailable_reason?: string | null
    selected: boolean
  }>
  selected_full_discount_id?: number | null
  selected_coupon_id?: number | null
  total_amount_cent: number
  discount_amount_cent: number
  full_discount_amount_cent: number
  coupon_discount_amount_cent: number
  points_discount_amount_cent: number
  points_used: number
  max_points_usable: number
  pay_amount_cent: number
}

export type OrderItem = {
  id: number
  product_id: number
  sku_id: number
  product_name: string
  sku_name: string
  cover_url?: string | null
  unit_price_cent: number
  quantity: number
  total_amount_cent: number
}

export type Order = {
  id: number
  order_no: string
  payment_id: number
  merchant_id: number
  merchant_name: string
  merchant_logo_url?: string | null
  status: string
  order_type?: string
  total_amount_cent: number
  pay_amount_cent: number
  full_discount_amount_cent: number
  coupon_discount_amount_cent: number
  points_discount_amount_cent: number
  points_used: number
  source_post_id?: number | null
  source_user_id?: number | null
  group_buy_activity_id?: number | null
  group_buy_group_id?: number | null
  shipping_address?: {
    receiver_name: string
    receiver_mobile: string
    province: string
    city: string
    district?: string | null
    street?: string | null
    detail_address: string
    postal_code?: string | null
    address_tag?: string | null
  } | null
  logistics_company?: string | null
  tracking_no?: string | null
  created_at?: string | null
  shipped_at?: string | null
  received_at?: string | null
  items: OrderItem[]
}

export type Payment = {
  id: number
  payment_no: string
  status: string
  pay_amount_cent: number
  points_used: number
  points_discount_amount_cent: number
  channel: string
  alipay_trade_no?: string | null
  alipay_qr_code?: string | null
  alipay_buyer_logon_id?: string | null
  order_ids: number[]
  paid_at?: string | null
}

export type AlipayPrecreateResult = {
  payment: Payment
  qr_code: string
  payment_no: string
  expire_minutes: number
}

export type Review = {
  id: number
  status: string
}

export type Refund = {
  id: number
  order_id: number
  order_item_id?: number | null
  product_id?: number | null
  sku_id?: number | null
  user_id: number
  quantity: number
  refund_amount_cent: number
  reason_type: string
  reason: string
  image_urls: string[]
  status: string
  origin_order_status: string
  created_at?: string | null
  updated_at?: string | null
  logs: {
    id: number
    operator_type: string
    operator_id?: number | null
    action: string
    message: string
    created_at?: string | null
  }[]
}

export type PageResult<T> = {
  list: T[]
  page: number
  page_size: number
  total: number
}

export const orderService = {
  listCart() {
    return http.get<unknown, { data: CartItem[] }>('/cart')
  },

  addCartItem(payload: { sku_id: number; quantity: number; source_post_id?: number | null }) {
    return http.post<unknown, { data: CartItem[] }>('/cart', payload)
  },

  updateCartItem(skuId: number, payload: { quantity: number; checked?: boolean }) {
    return http.put<unknown, { data: CartItem[] }>(`/cart/${skuId}`, {
      checked: true,
      ...payload,
    })
  },

  deleteCartItem(skuId: number) {
    return http.delete<unknown, { data: CartItem[] }>(`/cart/${skuId}`)
  },

  batchUpdateCartItems(payload: { sku_ids: number[]; checked: boolean }) {
    return http.patch<unknown, { data: CartItem[] }>('/cart/batch', payload)
  },

  batchDeleteCartItems(payload: { sku_ids?: number[] | null } = {}) {
    return http.delete<unknown, { data: CartItem[] }>('/cart', { data: payload })
  },

  checkout(payload: { full_discount_id?: number | null; coupon_id?: number | null; points_used?: number } = {}) {
    return http.post<unknown, { data: CheckoutResult }>('/cart/checkout', payload)
  },

  createOrder(payload: {
    client_order_token: string
    shipping_address_id?: number | null
    full_discount_id?: number | null
    coupon_id?: number | null
    points_used?: number
    source_post_id?: number | null
  }) {
    return http.post<unknown, { data: { payment_id: number; order_ids: number[]; pay_amount_cent: number } }>(
      '/orders',
      payload,
    )
  },

  precreateAlipay(paymentId: number, force = false) {
    return http.post<unknown, { data: AlipayPrecreateResult }>(`/payments/${paymentId}/alipay/precreate`, undefined, {
      params: { force },
    })
  },

  syncAlipay(paymentId: number) {
    return http.post<unknown, { data: Payment }>(`/payments/${paymentId}/alipay/sync`)
  },

  getPayment(paymentId: number) {
    return http.get<unknown, { data: Payment }>(`/payments/${paymentId}`)
  },

  listOrders(params?: { status?: string; page?: number; page_size?: number }) {
    return http.get<unknown, { data: PageResult<Order> }>('/orders', { params })
  },

  getOrder(orderId: number) {
    return http.get<unknown, { data: Order }>(`/orders/${orderId}`)
  },

  cancelOrder(orderId: number) {
    return http.post<unknown, { data: Order }>(`/orders/${orderId}/cancel`)
  },

  confirmOrder(orderId: number) {
    return http.post<unknown, { data: Order }>(`/orders/${orderId}/confirm`)
  },

  reviewOrder(orderId: number, payload: { product_id: number; score: number; content: string; image_urls?: string[] }) {
    return http.post<unknown, { data: Review }>(`/orders/${orderId}/reviews`, payload)
  },

  applyRefund(orderId: number, payload: { order_item_id: number; quantity: number; reason_type?: string; reason: string; image_urls?: string[] }) {
    return http.post<unknown, { data: Refund }>(`/orders/${orderId}/refunds`, payload)
  },

  listRefunds(params?: { status?: string }) {
    return http.get<unknown, { data: PageResult<Refund> }>('/orders/refunds', { params })
  },

  getRefund(refundId: number) {
    return http.get<unknown, { data: Refund }>(`/orders/refunds/${refundId}`)
  },
}

File: app/schemas/order.py
from pydantic import BaseModel, Field

from app.schemas.address import AddressResponse


class CartAddRequest(BaseModel):
    sku_id: int
    quantity: int = Field(default=1, ge=1)
    source_post_id: int | None = None


class CartUpdateRequest(BaseModel):
    quantity: int = Field(default=1, ge=1)
    checked: bool = True


class CartBatchUpdateRequest(BaseModel):
    sku_ids: list[int] = Field(min_length=1, max_length=100)
    checked: bool


class CartBatchDeleteRequest(BaseModel):
    sku_ids: list[int] | None = Field(default=None, max_length=100)


class CartItemResponse(BaseModel):
    sku_id: int
    product_id: int
    product_name: str
    sku_name: str
    price_cent: int
    quantity: int
    checked: bool
    cover_url: str | None = None
    source_post_id: int | None = None
    source_label: str | None = None
    invalid_reason: str | None = None


class CheckoutItemRequest(BaseModel):
    sku_id: int
    quantity: int = Field(ge=1)


class CheckoutRequest(BaseModel):
    items: list[CheckoutItemRequest] | None = None
    full_discount_id: int | None = None
    coupon_id: int | None = None
    points_used: int = Field(default=0, ge=0)


class PromotionOptionResponse(BaseModel):
    id: int
    name: str
    scope_type: str
    scope_ids: list[int] = Field(default_factory=list)
    min_amount_cent: int
    discount_amount_cent: int
    applicable_amount_cent: int
    available: bool = True
    unavailable_reason: str | None = None
    selected: bool = False


class CouponOptionResponse(BaseModel):
    id: int
    coupon_template_id: int
    name: str
    scope_type: str
    scope_ids: list[int] = Field(default_factory=list)
    discount_type: str
    discount_value: int
    min_amount_cent: int
    applicable_amount_cent: int
    discount_amount_cent: int
    status: str
    available: bool = True
    unavailable_reason: str | None = None
    selected: bool = False


class CheckoutResponse(BaseModel):
    items: list[CartItemResponse]
    addresses: list[AddressResponse] = Field(default_factory=list)
    available_full_discounts: list[PromotionOptionResponse] = Field(default_factory=list)
    available_coupons: list[CouponOptionResponse] = Field(default_factory=list)
    selected_full_discount_id: int | None = None
    selected_coupon_id: int | None = None
    total_amount_cent: int
    discount_amount_cent: int = 0
    full_discount_amount_cent: int = 0
    coupon_discount_amount_cent: int = 0
    points_discount_amount_cent: int = 0
    points_used: int = 0
    max_points_usable: int = 0
    pay_amount_cent: int


class CreateOrderRequest(BaseModel):
    client_order_token: str = Field(min_length=1, max_length=80)
    shipping_address_id: int | None = None
    full_discount_id: int | None = None
    coupon_id: int | None = None
    points_used: int = Field(default=0, ge=0)
    source_post_id: int | None = None
    items: list[CheckoutItemRequest] | None = None


class CreateOrderResponse(BaseModel):
    payment_id: int
    payment_no: str
    order_ids: list[int]
    pay_amount_cent: int
    expire_at: str | None = None


class ShippingAddressSnapshot(BaseModel):
    receiver_name: str
    receiver_mobile: str
    province: str
    city: str
    district: str | None = None
    street: str | None = None
    detail_address: str
    postal_code: str | None = None
    address_tag: str | None = None


class ShipOrderRequest(BaseModel):
    logistics_company: str = Field(min_length=1, max_length=80)
    tracking_no: str = Field(min_length=1, max_length=80)


class OrderItemResponse(BaseModel):
    id: int
    product_id: int
    sku_id: int
    product_name: str
    sku_name: str
    unit_price_cent: int
    quantity: int
    total_amount_cent: int

    model_config = {"from_attributes": True}


class OrderResponse(BaseModel):
    id: int
    order_no: str
    payment_id: int
    merchant_id: int
    status: str
    total_amount_cent: int
    pay_amount_cent: int
    full_discount_amount_cent: int = 0
    coupon_discount_amount_cent: int = 0
    points_discount_amount_cent: int = 0
    points_used: int = 0
    source_post_id: int | None = None
    source_user_id: int | None = None
    order_type: str = "normal"
    group_buy_activity_id: int | None = None
    group_buy_group_id: int | None = None
    shipping_address: ShippingAddressSnapshot | None = None
    logistics_company: str | None = None
    tracking_no: str | None = None
    shipped_at: str | None = None
    received_at: str | None = None
    items: list[OrderItemResponse]

    model_config = {"from_attributes": True}


class PaymentResponse(BaseModel):
    id: int
    payment_no: str
    status: str
    pay_amount_cent: int
    points_used: int = 0
    points_discount_amount_cent: int = 0
    channel: str = "mock"
    alipay_trade_no: str | None = None
    alipay_qr_code: str | None = None
    alipay_buyer_logon_id: str | None = None
    order_ids: list[int] = Field(default_factory=list)
    paid_at: str | None = None

    model_config = {"from_attributes": True}


class AlipayPrecreateResponse(BaseModel):
    payment: PaymentResponse
    qr_code: str
    payment_no: str
    expire_minutes: int


class AlipayNotifyResponse(BaseModel):
    success: bool


class ReviewCreateRequest(BaseModel):
    product_id: int
    score: int = Field(ge=1, le=5)
    content: str = Field(default="", max_length=1000)
    image_urls: list[str] = Field(default_factory=list)


class ReviewAuditRequest(BaseModel):
    approved: bool


class ReviewResponse(BaseModel):
    id: int
    user_id: int
    user_nickname: str | None = None
    user_avatar_url: str | None = None
    order_id: int
    product_id: int
    score: int
    content: str
    image_urls: list[str]
    status: str


class RefundCreateRequest(BaseModel):
    order_item_id: int
    quantity: int = Field(default=1, ge=1)
    reason_type: str = Field(default="other", max_length=50)
    reason: str = Field(min_length=1, max_length=255)
    image_urls: list[str] = Field(default_factory=list)


class RefundLogResponse(BaseModel):
    id: int
    operator_type: str
    operator_id: int | None = None
    action: str
    message: str
    created_at: str | None = None


class RefundResponse(BaseModel):
    id: int
    order_id: int
    order_item_id: int | None = None
    product_id: int | None = None
    sku_id: int | None = None
    user_id: int
    quantity: int = 0
    refund_amount_cent: int
    reason_type: str
    reason: str
    image_urls: list[str] = Field(default_factory=list)
    status: str
    origin_order_status: str
    created_at: str | None = None
    updated_at: str | None = None
    logs: list[RefundLogResponse] = Field(default_factory=list)

    model_config = {"from_attributes": True}
